lib/FIR/statements.h

```diff
@@ -592,6 +580,14 @@ class Statement : public SumTypeMixin<std::variant<ReturnStmt,  //
     parent->insertBefore(this);
   }
   std::string dump() const;
+
+  static std::size_t offsetof_impl() {
+    return reinterpret_cast<std::size_t>(&static_cast<Statement *>(nullptr)->u);
+  }
+  static Statement *From(Stmt_impl *stmt) {
+    return reinterpret_cast<Statement *>(
+        reinterpret_cast<char *>(stmt) - Statement::offsetof_impl());
```

Comment:
This has me baffled. Can you add a comment about what's going on and why it's safe?

Reply:
This may not be the way to go.  I'll make an analogy to explain this particular implementation however.
Say we have a world of envelopes and contents.  For some purposes, it is great to deal with envelopes without regard to the letter inside them.  For other purposes, it is better to know what is inside the envelope. (Say maybe knowing which are bills, which are checks, etc.) All this was intended to be was a quick way to map from "this electric bill" back to the envelope it is packaged in. This should work fine in a world where we know all content is wrapped in its own envelope.
Unfortunately, C++ doesn't have a robust and cheap way of encoding this relationship apparently.
For your amusement, here's a link to a more elaborate "envelope finder" scheme in LLVM. http://llvm.org/docs/ProgrammersManual.html#the-waymarking-algorithm